### PR TITLE
DAOS-14010 container: maintain stable epoch

### PR DESCRIPTION
**Pull request by wangdi1**
Maintain stable epoch on each target as the earliest open dtx epoch - 1, which will be reported to the PS leader, which will choose the minimum one from all target, then sync it to all targets as the VOS aggregation upper boundary, and also as the stable epoch of each target, which will be used by incremental reintegration as the start epoch to reconstruct the data.

Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: src/vos/vos_layout.h [hunk 1]
```diff
@@ -246,12 +246,15 @@ enum vos_io_stream {
...
 
+#define DAOS_CONT_V1	0x0001
+#define DAOS_CONT_VERSION	DAOS_CONT_V1
 /* VOS Container Value */
 struct vos_cont_df {
 	uuid_t				cd_id;
 	uint64_t			cd_nobjs;
 	uint32_t			cd_ts_idx;
-	uint32_t			cd_pad;
+	uint16_t			cd_ver;
```
Comment (jolivier23) [inline src/vos/vos_layout.h:256]: We already maintain the POOL_DF version in pd_version field.   This is updated on pool upgrade.  Do we need to support more than one container version at a time?  Can we just iterate containers in the pool in vos_pool_upgrade and reallocate each container root so as to maintain a single container version per pool?

Comment (jolivier23) [reply src/vos/vos_layout.h:256]: The idea here would be to control this using pool->vp_feats which is setup on initialization based on the pool_df.

Comment (wangdi1) [reply src/vos/vos_layout.h:256]: sure, will update the patch. Thanks.

**Comment by wangdi1** [comment]
merged this patch to https://github.com/daos-stack/daos/pull/13109

**Pull request closed without merging**